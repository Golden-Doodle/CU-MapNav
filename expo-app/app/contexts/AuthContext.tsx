// contexts/AuthContext.tsx
import React, { createContext, useState, useEffect } from "react";
import { Alert } from "react-native"; // <-- Import Alert
import AsyncStorage from "@react-native-async-storage/async-storage";
import { GoogleSignin } from "@react-native-google-signin/google-signin";
import auth, { FirebaseAuthTypes } from "@react-native-firebase/auth";
import { useRouter } from "expo-router";
import {
  fetchAllCalendars,
  fetchGoogleCalendarEvents,
} from "@/app/services/GoogleCalendar/fetchingUserCalendarData";
import { GoogleCalendarEvent } from "../utils/types";

export interface AuthContextType {
  user: FirebaseAuthTypes.User | null;
  setUser: React.Dispatch<React.SetStateAction<FirebaseAuthTypes.User | null>>;
  loading: boolean;
  handleGoogleSignIn: () => Promise<void>;
  signOut: () => Promise<void>;
  handleSignInAsGuest: () => Promise<void>;
  googleCalendarEvents: GoogleCalendarEvent[];
  selectedCalendarId: string | null;
  setSelectedCalendarId: (id: string) => void;
}

export const AuthContext = createContext<AuthContextType | null>(null);

export const AuthProvider = ({ children }: { children: React.ReactNode }) => {
  const router = useRouter();
  const [user, setUser] = useState<FirebaseAuthTypes.User | null>(null);
  const [loading, setLoading] = useState(true);
  const [googleCalendarEvents, setGoogleCalendarEvents] = useState<GoogleCalendarEvent[]>([]);
  const [selectedCalendarId, setSelectedCalendarId] = useState<string | null>(null);

  useEffect(() => {
    GoogleSignin.configure({
      webClientId: "259837654437-eo18pu30v9grv1i3dog8ba5i64ipj1q7.apps.googleusercontent.com",
      scopes: ["https://www.googleapis.com/auth/calendar.readonly"],
    });
  }, []);

  useEffect(() => {
    const loadUser = async () => {
      try {
        const storedUser = await AsyncStorage.getItem("user");
        setUser(storedUser ? JSON.parse(storedUser) : null);
      } catch (error) {
        console.error("Error loading user:", error);
      }
      setLoading(false);
    };
    loadUser();
  }, []);

  useEffect(() => {
    const unsubscribe = auth().onAuthStateChanged(async (firebaseUser) => {
      setUser(firebaseUser);
      if (firebaseUser) {
        await AsyncStorage.setItem("user", JSON.stringify(firebaseUser));
      } else {
        await AsyncStorage.removeItem("user");
      }
    });

    return () => unsubscribe();
  }, []);

  useEffect(() => {
    if (!user) return;

    const fetchData = async () => {
      try {
        let storedCalendarId = await AsyncStorage.getItem("selectedScheduleID");

        if (!storedCalendarId) {
          const allCalendars = await fetchAllCalendars();
          const defaultCalendar = allCalendars.find(
            (cal: any) => cal.summary === "Concordia_Class_Schedule"
          );

          if (defaultCalendar) {
            storedCalendarId = defaultCalendar.id;
            await AsyncStorage.setItem("selectedScheduleID", defaultCalendar.id);
            await AsyncStorage.setItem("selectedScheduleName", defaultCalendar.summary);
          } else {
            console.warn("Default calendar 'Concordia_Class_Schedule' not found.");
          }
        }

        if (storedCalendarId) {
          setSelectedCalendarId(storedCalendarId);

          // Fetch events and catch UNAUTHORIZED error
          const events = await fetchGoogleCalendarEvents(storedCalendarId, 7);
          setGoogleCalendarEvents(events);
        }
      } catch (error) {
        // If we detect the "UNAUTHORIZED" error, log out user and show an alert
        if ((error as Error).message === "UNAUTHORIZED") {
          Alert.alert(
            "Session Expired",
            "Your Google session has expired. Please sign in again.",
            [
              {
                text: "OK",
                onPress: () => signOut(),
              },
            ]
          );
        } else {
          console.error("Error fetching calendar events:", error);
        }
      }
    };

    fetchData();
  }, [user]);

  const handleGoogleSignIn = async () => {
    try {
      await GoogleSignin.hasPlayServices({ showPlayServicesUpdateDialog: true });
      const signInResponse = await GoogleSignin.signIn();
      const { idToken, accessToken } = await GoogleSignin.getTokens();

      if (!idToken) throw new Error("Google Sign-In failed: No ID Token received.");
      if (!accessToken) throw new Error("Google Sign-In failed: No Access Token received.");

      const googleCredential = auth.GoogleAuthProvider.credential(idToken);
      const userCredential = await auth().signInWithCredential(googleCredential);

      await AsyncStorage.setItem("user", JSON.stringify(userCredential.user));
      await AsyncStorage.setItem("googleAccessToken", accessToken);

      router.replace("/screens/Home/HomePageScreen");
    } catch (error) {
      console.error("Google Sign-In Error:", error);
    }
  };

  const handleSignInAsGuest = async () => {
    router.replace("/screens/Home/HomePageScreen");
  };

  const signOut = async () => {
    try {
      await GoogleSignin.revokeAccess();
      await GoogleSignin.signOut();
      await auth().signOut();
      await AsyncStorage.removeItem("user");
      await AsyncStorage.removeItem("selectedScheduleID");
      await AsyncStorage.removeItem("selectedScheduleName");
      await AsyncStorage.removeItem("googleAccessToken");
      setUser(null);
      setSelectedCalendarId(null);
      setGoogleCalendarEvents([]);
      router.replace("/");
    } catch (error) {
      console.error("Logout Error:", error);
    }
  };

  return (
    <AuthContext.Provider
      value={{
        user,
        setUser,
        loading,
        handleGoogleSignIn,
        signOut,
        handleSignInAsGuest,
        googleCalendarEvents,
        selectedCalendarId,
        setSelectedCalendarId,
      }}
    >
      {children}
    </AuthContext.Provider>
  );
};
