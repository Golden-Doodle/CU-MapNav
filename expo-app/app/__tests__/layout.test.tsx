import React from "react";
import { render, waitFor } from "@testing-library/react-native";
import RootLayout from "../_layout";
import { AuthProvider } from "@/app/contexts/AuthContext";
import { Stack } from "expo-router";

jest.mock("expo-router", () => ({
  Stack: jest.fn(() => null),
  useRouter: jest.fn(() => ({ push: jest.fn() })),
}));

jest.mock("@react-native-google-signin/google-signin", () => ({
  GoogleSignin: {
    configure: jest.fn(),
    hasPlayServices: jest.fn().mockResolvedValue(true),
    signIn: jest.fn().mockResolvedValue({ idToken: "test-id-token" }),
    getTokens: jest.fn().mockResolvedValue({ idToken: "test-id-token", accessToken: "test-access-token" }),
    revokeAccess: jest.fn(),
    signOut: jest.fn(),
  },
}));

jest.mock("@react-native-firebase/auth", () => {
  return () => ({
    onAuthStateChanged: jest.fn((callback) => {
      callback(null);
      return jest.fn();
    }),
    signInWithCredential: jest.fn().mockResolvedValue({
      user: { uid: "1", email: "test@example.com" },
    }),
    signOut: jest.fn(),
    currentUser: null,
  });
});

jest.mock("instabug-reactnative", () => ({
  init: jest.fn(),
  setPrimaryColor: jest.fn(),
  InvocationEvent: {
    shake: "shake",
    screenshot: "screenshot",
    floatingButton: "floatingButton",
  },
}));


describe("RootLayout", () => {
  it("renders the AuthProvider and Stack navigator correctly", async () => {
    await waitFor(() => {
      const { toJSON } = render(<RootLayout />);
      expect(toJSON()).toMatchSnapshot();
    });
  });

  it("wraps the application with AuthProvider", async () => {
    await waitFor(() => {
      const { UNSAFE_getByType } = render(<RootLayout />);
      expect(UNSAFE_getByType(AuthProvider)).toBeTruthy();
    });
  });

  it("renders the Stack navigator with headerHidden option", async () => {
    await waitFor(() => {
      render(<RootLayout />);
      expect(Stack).toHaveBeenCalledWith(
        { screenOptions: { headerShown: false } },
        {}
      );
    });
  });
});
