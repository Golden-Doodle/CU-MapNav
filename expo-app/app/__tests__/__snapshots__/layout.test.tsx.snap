// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RootLayout renders the AuthProvider and Stack navigator correctly 1`] = `null`;
